File: 프로그래머스/4/60060. 가사 검색/가사 검색.java
import java.util.*;

class Solution {
    public int[] solution(String[] words, String[] queries) {                

        Node trie = new Node();
        Node reverseTrie = new Node();

        for (String word : words) {
            trie.add(word, 0);
            reverseTrie.add(new StringBuilder(word).reverse().toString(), 0);
        }
        
        int n = queries.length;
        
        int[] ans = new int[n];

        for (int i = 0; i < n; i++) {
            String query = queries[i];

            ans[i] = query.startsWith("?")
                    ? reverseTrie.count(new StringBuilder(query).reverse().toString(), 0)
                    : trie.count(query, 0);
        }

        return ans;
    }
    
    static class Node {

        Map<Integer, Integer> countMap = new HashMap<>();
        Map<Character, Node> children = new HashMap<>();

        public void add(String word, int index) {

            if (index >= word.length()) return;

            int length = word.length() - index;
            countMap.put(length, countMap.getOrDefault(length, 0) + 1);

            char c = word.charAt(index);
            Node child = children.getOrDefault(c, new Node());

            child.add(word, index + 1);

            children.put(c, child);
        }

        public int count(String query, int index) {

            if (query.charAt(index) == '?') {
                return countMap.getOrDefault(query.length() - index, 0);
            }

            char c = query.charAt(index);

            if (!children.containsKey(c)) return 0;

            Node child = children.get(c);
            return child.count(query, index + 1);
        }
    }
}

File: 프로그래머스/4/60060. 가사 검색/가사 검색.py
from collections import defaultdict
import sys
sys.setrecursionlimit(10**6)

class Node:
    def __init__(self):
        self.count_map = defaultdict(int)
        self.children = {}

    def add(self, word, index):
        if index >= len(word):
            return

        length = len(word) - index
        self.count_map[length] += 1

        c = word[index]
        if c not in self.children:
            self.children[c] = Node()

        self.children[c].add(word, index + 1)

    def count(self, query, index):
        if query[index] == '?':
            return self.count_map[len(query) - index]

        c = query[index]
        if c not in self.children:
            return 0

        return self.children[c].count(query, index + 1)

def solution(words, queries):
    
    trie = Node()
    reverseTrie = Node()

    for word in words:
        trie.add(word, 0)
        reverseTrie.add(word[::-1], 0)

    ans = []
    for query in queries:
        if query[0] == '?':
            ans.append(reverseTrie.count(query[::-1], 0))            
        else:
            ans.append(trie.count(query, 0))

    return ans